File: go/300.500/472.ConcatenatedWords.go
package main

import (
	"fmt"
	"sort"
)

type Trie struct {
	IsEnd    bool
	Children []*Trie
}

func NewTrie() *Trie {
	return &Trie{
		IsEnd:    false,
		Children: make([]*Trie, 26),
	}
}

func findAllConcatenatedWordsInADict(words []string) []string {
	sort.Slice(words, func(i, j int) bool {
		return len(words[i]) < len(words[j])
	})
	head := NewTrie()
	ans := []string{}
	for _, word := range words {
		visited := make([]bool, len(word))
		if dfs(head, word, 0, visited) {
			ans = append(ans, word)
		} else {
			insert(head, word)
		}
	}
	return ans
}

func dfs(head *Trie, word string, start int, visited []bool) bool {
	size := len(word)
	if size == start {
		return true
	}
	if visited[start] {
		return false
	}
	visited[start] = true
	node := head
	for i := start; i < size; i++ {
		index := int(word[i] - 'a')
		node = node.Children[index]
		if node == nil {
			return false
		}
		if node.IsEnd {
			if dfs(head, word, i+1, visited) {
				return true
			}
		}
	}
	return false
}

func insert(node *Trie, word string) {
	for _, c := range word {
		index := int(c - 'a')
		child := node.Children[index]
		if child == nil {
			child = NewTrie()
			node.Children[index] = child
		}
		node = child
	}
	node.IsEnd = true
}

func findAllConcatenatedWordsInADict1(words []string) []string {
	dic := map[string]struct{}{}
	for i := range words {
		dic[words[i]] = struct{}{}
	}
	ret := []string{}
	check := func(x int) {
		size := len(words[x])
		dp := make([]int, size+1)
		for i := 1; i <= size; i++ {
			dp[i] = -1
		}
		for i := 0; i <= size; i++ {
			if dp[i] < 0 {
				continue //表示 word[0..i]不再dic里面,没必要进行后续判断
			}
			for j := i + 1; j <= size; j++ {
				if _, ok := dic[words[x][i:j]]; ok {
					//word[i..j]也在dic里面，可以从j开始继续判断
					dp[j] = max(dp[j], dp[i]+1)
				}
			}
			if dp[size] > 1 { //表示中间至少有一个分段
				ret = append(ret, words[x])
				return
			}
		}
	}
	for x := range words {
		check(x)
	}
	return ret
}

func max(i, j int) int {
	if i > j {
		return i
	}
	return j
}

func main() {
	fmt.Println(findAllConcatenatedWordsInADict1([]string{"cat", "cats", "catsdogcats", "dog", "dogcatsdog", "hippopotamuses", "rat", "ratcatdogcat"}), `["catsdogcats","dogcatsdog","ratcatdogcat"]`)
	fmt.Println(findAllConcatenatedWordsInADict1([]string{"cat", "dog", "catdog"}), `["catdog"]`)
}

File: 300.500/472.ConcatenatedWords.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
#include <unordered_set>
using namespace std;

struct Trie{
    bool isEnd{false};
    vector<Trie*> children = vector<Trie*>(26, nullptr);
};

class ConcatenatedWords{
public:
    vector<string> findAllConcatenatedWordsInADict(vector<string> words) {
        vector<string> ans;
        //先按长度排序 保证处理较长字符串时 比他短的字符串都处理过了
        sort(words.begin(), words.end(), [](const string& a, const string& b){
            return a.size() < b.size();
        });
        for(auto& word : words){
            if(word.size() == 0){
                continue;
            }
            //判断字符串是否是连接词
            vector<bool> visited(word.size(), false);
            if(dfs(word, 0, visited)){
                //如果是连接词 加入答案
                ans.push_back(word);
            }else{
                //如果不是 加入字典树
                insert(word);
            }
        }
        return ans;
    }
private:
    bool dfs(const string& word, int start, vector<bool>& visited){
        if(word.size() == start){
            //已经遍历完整word了
            return true;
        }
        if(visited[start] == true){
            //从这里开始已经检测过了 所以不需要再检测了
            return false;
        }
        visited[start] = true;
        Trie* node = &trie;
        for(int i = start;i<word.size();++i){
            char ch = word[i];
            int index = ch -'a';
            node = node->children[index];
            if(node == nullptr){
                return false;
            }
            if(node->isEnd){
                //从这里分词试试
                if(dfs(word, i+1, visited)){
                    return true;
                }
            }
        }
        return false;
    } 
    void insert(string& word){
        Trie* node = &trie;
        for(int i =0;i<word.size();++i){
            char ch = word[i];
            int index = ch - 'a';
            if(node->children[index] == nullptr){
                node->children[index] = new Trie{};
            }
            node = node->children[index];
        }
        node->isEnd = true;
    }
private:
    Trie trie;
};

string print(vector<string> data){
    string str = "[";
    for (auto& s : data){
        str += s + ",";
    }
    str.pop_back();
    str += "]";
    return str;
}

class ConcatenatedWords1{
public:
    vector<string> findAllConcatenatedWordsInADict(vector<string> words) {
        for(auto& word : words){
            dic.insert(word);
        }
        vector<string> ret;
        for(auto &word : words){
            if(check(word)){
                ret.push_back(word);
            }
        }
        return ret;
    }
private:
    bool check(string& word){
        int size = word.size();
        vector<int> dp(size+1, -1);
        dp[0] = 0;
        for(int i =0;i<size;++i){
            if(dp[i] < 0){
                continue;
            }
            for(int j =i+1;j<=size;++j){
                if(dic.count(word.substr(i, j-i)) > 0){
                    dp[j] = max(dp[j], dp[i] + 1);
                }
            }
            if(dp[size] > 1){
                return true;
            }
        }
        return false;
    }
private:
    unordered_set<string> dic;
};

int main(){
    ConcatenatedWords1 cls;
    cout << print(cls.findAllConcatenatedWordsInADict(vector<string>{"cat","cats","catsdogcats","dog","dogcatsdog","hippopotamuses","rat","ratcatdogcat"})).c_str() << " -> [catsdogcats,dogcatsdog,ratcatdogcat]" << endl;
    ConcatenatedWords1 cls1;
    cout << print(cls1.findAllConcatenatedWordsInADict(vector<string>{"cat","dog","catdog"})).c_str() << " -> [catdog]" << endl;
    return 0;
}
